### ISSUE TITLE
Clarification NER Training Volume and Ratio Mixing-In 'Examples of other Entity Types'

### ISSUE THREAD

**Issue by moritz-volland** [explosion/spaCy#1054]
Hi,

First of all, thanks for doing spaCy it's pretty cool. I'm in the process of (re-) training spaCy's Named Entity Recognizer and I have a couple of questions that I couldn't deduce from the examples provided and I would be very grateful if I could get some guidance.

 1. If a few hundred examples are considered 'a good starting point', then what would be a reasonable number to aim for? Is 100 000 entity/label excessive?
 2. If I introduce a new label, is it best if the number of the entities of that labelled are roughly the same (balanced) during training?
 3. Regarding the mixing in 'examples of other entity types':
    - do I just add random known categories/labels to my training set eg: `('The Business Standard published in its recent issue on crude oil and natural gas ...', [(4,21, 'ORG')], )`?

    - can I use the same text for various labels? e.g. `('The Business Standard published in its recent issue on crude oil and natural gas ...', [(55,64, 'COMMODITY')], )`?

- on a similar note let's assume I want spaCy's NER to also recognize a second `COMMODITY` could I then also just use the same sentence and label a different region e.g. `('The Business Standard published in its recent issue on crude oil and natural gas ...', [(69,80, 'COMMODITY')], )`? Is that how it's supposed to be done?

    - what ratio between new and other (old) labels is considered reasonable

###Thanks####

## My Environment
* Python2.7
* Ubuntu 16.04
* spaCy 1.8.2

PS This question has also been posted on http://stackoverflow.com/questions/43912195/re-training-spacys-ner-training-volume-and-mix-of-entity-types

**Comment by macarthy** [comment]
+1

**Comment by honnibal** [comment]
These are all very good questions, but it's hard to give very satisfying answers. Each machine learning problem currently requires individual attention, because we don't have one-size-fits-all recipes for these things. General advice about machine learning applies -- you might try Andrew Ng's primer "Machine Learning Yearning". I only saw sample chapters, but it looked like a good primer on these practical issues.

The only definitive answer I can give is about giving annotations on the same sentence: spaCy assumes that when the annotations you set include *all* the entities in that text --- so it will mark words outside the entities as tag `O`. This may be incorrect if you just want to say "I know this word is an entity, but I don't know anything about the others". If you have incomplete annotations, a good solution is to load two copies of the model: the original, and one you'll be updating. Use the original model to fill in missing entities, for words where you don't know the annotation. This keeps the model you're training from diverging too far in its predictions from the original state.

> If a few hundred examples are considered 'a good starting point', then what would be a reasonable number to aim for? Is 100 000 entity/label excessive?

Every machine learning problem will have a different examples/accuracy curve. You can get an idea for this by training with less data than you have, and seeing what the curve looks like. If you have 1,000 examples, then try training with 500, 750, etc, and see how that affects your accuracy.

> If I introduce a new label, is it best if the number of the entities of that labelled are roughly the same (balanced) during training?

There's  trade-off between making the gradients too sparse, and making the learning problem too unrepresentative of what the actual examples will look like.

**Comment by moritz-volland** [comment]
I very much appreciate the comprehensive answer. Thanks for taking the time.

PS from my point of view the issue can be closed

**Comment by ankur7721** [comment]
@honnibal @moritz-volland @macarthy I had the same doubts and followed the suggestion.

Here's what I tried as described [here]
 (https://github.com/explosion/spaCy/blob/master/examples/training/train_ner.py), where I am trying to add more examples to entity `MONEY`. I know the training examples are only two but it shouldn't have given all label as `MONEY`.

```
from __future__ import unicode_literals, print_function
import json
import pathlib
import random
import re
import spacy
from spacy.pipeline import EntityRecognizer
from spacy.gold import GoldParse
from spacy.tagger import Tagger

try:
    unicode
except:
    unicode = str

def train_ner(nlp, train_data, entity_types):
    for raw_text, _ in train_data:
        doc = nlp.make_doc(raw_text)
        for word in doc:
            _ = nlp.vocab[word.orth]

    # Train NER.
    ner = EntityRecognizer(nlp.vocab, entity_types=entity_types)
    for itn in range(5):
        random.shuffle(train_data)
        for raw_text, entity_offsets in train_data:
            doc = nlp.make_doc(raw_text)
            gold = GoldParse(doc, entities=entity_offsets)
            ner.update(doc, gold)
    ner.model.end_training()
    return ner

def save_model(ner, model_dir):
    model_dir = pathlib.Path(model_dir)
    if not model_dir.exists():
        model_dir.mkdir()
    assert model_dir.is_dir()

    with (model_dir / 'config.json').open('wb') as file_:
        data = json.dumps(ner.cfg)
        if isinstance(data, unicode):
            data = data.encode('utf8')
        file_.write(data)
    ner.model.dump(str(model_dir / 'model'))
    if not (model_dir / 'vocab').exists():
        (model_dir / 'vocab').mkdir()
    ner.vocab.dump(str(model_dir / 'vocab' / 'lexemes.bin'))
    with (model_dir / 'vocab' / 'strings.json').open('w', encoding='utf8') as file_:
        ner.vocab.strings.dump(file_)

def cleanup(token, lower=True):
    if lower:
        token = token.lower()
    return token.strip()


def main(model_dir=None):
    nlp = spacy.load('en_core_web_md', parser=False, entity=False, add_vectors=False)

    if nlp.tagger is None:
        print('---- WARNING ----')
        print('Data directory not found')
        print('please run: `python -m spacy.en.download --force all` for better performance')
        print('Using feature templates for tagging')
        print('-----------------')
        nlp.tagger = Tagger(nlp.vocab, features=Tagger.feature_templates)

    train_data = [
        (
            'Total Current Charges 369.73',
            [(len('Total Current Charges '), len('Total Current Charges 369.73'),
              'MONEY')]
        ),
        (
            'Amount paid Rs.350.0',
            [(len('Amount paid '), len('Amount paid Rs.350.0'),
              'MONEY')]
        )
    ]
    ner = train_ner(nlp, train_data, ['MONEY'])

    doc_text = unicode(open('input.txt').read().decode('utf8'))
    doc = nlp.make_doc(doc_text)
    nlp.tagger(doc)
    ner(doc)
    for ent in doc.ents:
        print(ent.label_, ent.text)

    if model_dir is not None:
        save_model(ner, model_dir)


if __name__ == '__main__':
    main('ner')

```


My text file input is 

`Total Amount ₹30 Any discrepancies? Contact us. 6/19/2017 Gmail - You've paid Rs 30 to Sree saravana catering https://mail.google.com/mail/u/`

while my output is

```
MONEY ₹30
MONEY discrepancies?
MONEY  
MONEY us
MONEY Gmail
MONEY You
MONEY Rs
MONEY 30
MONEY Sree
MONEY catering https://mail.google.com/mail/u/
```

which shouldn't be the case. Also, everytime I ran this, I get different set of results. What can i do to counter this. Instead of initializing with random weights, can I initialize with uniform weights [Ref](https://arxiv.org/pdf/1704.08863.pdf) ?
